### PR TITLE
aws - iam-profile, ec2 - add has-specific-managed-policy filter

### PR DESCRIPTION
**Pull request by naohito-intuit**
Following up on #7947, add the `has-specific-managed-policy` filter at the `iam-profile` resource and then reuse the code in the `ec2` resource.

Examples
```yaml
policies:
  - name: iam-profiles-have-admin-policy
    resource: iam-profile
    filters:
      - type: has-specific-managed-policy
        value: admin-policy
```
```yaml
policies:
  - name: ec2-instances-have-admin-policy
    resource: ec2
    filters:
      - type: has-specific-managed-policy
        value: admin-policy
```

### CODE REVIEW

**Comment by naohito-intuit** [comment]
@thisisshi please review.

**Review by thisisshi — APPROVED** [review]
thanks this is much better lgtm

File: c7n/resources/ec2.py [hunk 1]
```diff
@@ -2345,3 +2345,49 @@ def set_resource_desired(self, resource, desired):
...
+    def process(self, resources, event=None):
+        client = utils.local_session(self.manager.session_factory).client('iam')
+        iam_profiles = self.manager.get_resource_manager('iam-profile').resources()
+        iam_profiles_mapping = {profile['Arn']: profile for profile in iam_profiles}
+
+        results = []
+        for r in resources:
+            if r['State']['Name'] == 'terminated':
+                continue
+            instanceProfileArn = r.get('IamInstanceProfile', {}).get('Arn')
```
Comment (thisisshi) [inline c7n/resources/ec2.py:2382]:
```suggestion
            instance_profile_arn = r.get('IamInstanceProfile', {}).get('Arn')
```

File: c7n/resources/ec2.py [hunk 2]
```diff
@@ -2345,3 +2345,49 @@ def set_resource_desired(self, resource, desired):
...
+            if not instanceProfileArn:
```
Comment (thisisshi) [inline c7n/resources/ec2.py:2383]:
```suggestion
            if not instance_profile_arn:
```

File: c7n/resources/ec2.py [hunk 3]
```diff
@@ -2345,3 +2345,49 @@ def set_resource_desired(self, resource, desired):
...
+                continue
+
+            profile = iam_profiles_mapping.get(instanceProfileArn)
```
Comment (thisisshi) [inline c7n/resources/ec2.py:2386]:
```suggestion
            profile = iam_profiles_mapping.get(instance_profile_arn)
```

**Pull request merged**